src/Simulation/Common/IQuantumExecutor.cs

```diff
@@ -0,0 +1,583 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Quantum.Simulation.Core;
+
+namespace Microsoft.Quantum.Simulation.QuantumExecutor
+{
+    /// <summary>
+    /// An interface for implementing QDK target machines that work on a quantum circuit level. 
+    /// It is intended to be used with <see cref="QuantumExecutor"/>.
+    /// </summary>
+    /// <remarks>
+    /// Simulators implemented using <see cref="IQuantumExecutor"/> interface do not manage qubits on their own.
+    /// Instead they are notified when qubits are allocated, released, borrowed and returned.
+    /// </remarks>
+    public interface IQuantumExecutor
+    {
+        /// <summary>
+        /// Called when <a href="https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.intrinsic.x">Microsoft.Quantum.Intrinsic.X</a> is called in Q#.
+        /// In Q# the operation applies X gate to <paramref name="qubit"/>. The gate is given by matrix X=((0,1),(1,0)).
+        /// </summary>
+        /// <remarks>
+        /// When adjoint of X is called in Q#, <see cref="X(Qubit)"/> is called because X is self-adjoint.
+        /// The names and the order of the parameters are the same as for the corresponding Q# operation.
+        /// </remarks>
+        /// <param name="qubit">Qubit to which the gate should be applied.</param>
+        void X(Qubit qubit);
+
+        /// <summary>
+        /// Called when controlled <a href="https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.intrinsic.x">Microsoft.Quantum.Intrinsic.X</a> is called in Q#.
+        /// In Q# the operation applies X gate to <paramref name="qubit"/> controlled on <paramref name="controls"/>. The gate is given by matrix X=((0,1),(1,0)).
```

Comment:
The matrix is correct. It says that this matrix is applied to the target qubit, not to the tuple of qubits that include controls. Since there could be multiple controls, it is not possible to show matrix for all the qubits.

---
In reply to: [348329787](https://github.com/microsoft/qsharp-runtime/pull/82#discussion_r348329787) [](ancestors = 348329787)